File: DesignPattern/other_patterns/semaphore_pattern.py
import threading
import time

# Initialize a semaphore with a maximum of 2 permits
semaphore = threading.Semaphore(2)

def worker(thread_id):
    print(f"Thread {thread_id} is trying to acquire a permit.")
    
    with semaphore:
        print(f"Thread {thread_id} has acquired a permit and entered the critical section.")
        time.sleep(2)
        print(f"Thread {thread_id} has released the permit and exited the critical section.")

# Create and start multiple threads
threads = []
for i in range(4):
    thread = threading.Thread(target=worker, args=(i,))
    threads.append(thread)
    thread.start()

# Wait for all threads to finish
for thread in threads:
    thread.join()

print("All threads have finished.")



'''
- In Python, you can use the threading module to implement a semaphore pattern to control access to a shared resource in a multi-threaded environment. 
Here's an example of using the threading.Semaphore class to achieve semaphore-based synchronization:

- In this Python example: We use the threading.Semaphore class to create a semaphore with a maximum of 2 permits. This means that up to two threads 
can enter the critical section concurrently.

- The worker function simulates the work done by each thread. It attempts to acquire a permit from the semaphore using a with statement. When a permit 
is acquired, the thread enters the critical section, simulates some work (sleeps for 2 seconds), and then releases the permit when it exits the critical section.

- We create and start four threads that call the worker function.

- After starting the threads, we wait for all of them to finish by calling thread.join() for each thread.

- When you run this code, you'll see that at most two threads enter the critical section concurrently because the semaphore allows only two permits. The other 
threads wait until a permit is released, and they can acquire it to enter the critical section.

- The threading.Semaphore class provides a convenient way to implement semaphore-based synchronization in Python for managing access to shared 
resources among multiple threads.
'''

File: DesignPattern/other_patterns/semaphore_pattern.cpp
#include <iostream>
#include <thread>
#include <semaphore>

std::counting_semaphore<int> semaphore(2); // Allow up to 2 threads to enter the critical section

void worker(int id) {
    std::cout << "Thread " << id << " is waiting." << std::endl;
    semaphore.acquire(); // Try to acquire a permit

    // Critical section
    std::cout << "Thread " << id << " entered the critical section." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(2));
    std::cout << "Thread " << id << " exited the critical section." << std::endl;

    semaphore.release(); // Release the permit
}

int main() {
    std::thread t1(worker, 1);
    std::thread t2(worker, 2);
    std::thread t3(worker, 3);

    t1.join();
    t2.join();
    t3.join();

    return 0;
}



/*
- The Semaphore pattern is a synchronization primitive used to control access to a shared resource in a multi-threaded or multi-process environment. 
It helps in managing concurrent access to resources by allowing a limited number of threads or processes to enter a critical section simultaneously. 
In C++, you can use the std::semaphore introduced in C++20 to implement semaphore-based synchronization. Here's an example of using std::semaphore:

- In this example: We include the necessary headers for working with std::semaphore.

- We create a counting semaphore semaphore with an initial count of 2, which allows up to 2 threads to enter the critical section concurrently.

- The worker function represents the work that each thread does. It starts by waiting for the semaphore to acquire a permit.

- In the critical section, each thread prints a message, simulates work by sleeping for 2 seconds, and then prints another message.

- After exiting the critical section, the thread releases the permit by calling semaphore.release().

- In the main function, we create three threads (t1, t2, and t3) to demonstrate the semaphore-based synchronization. You can see that at most two 
threads can enter the critical section concurrently due to the semaphore's initial count of 2.

- Compile this code with C++20 support to use std::semaphore. This example shows how you can use semaphores to control concurrent access to a critical 
section in a multi-threaded C++ program.
*/
